src/kv/mod.rs

```diff
@@ -11,16 +11,228 @@
+//!
+//! impl<'kvs> VisitSource<'kvs> for Collect<'kvs> {
+//!     fn visit_pair(&mut self, key: Key<'kvs>, value: Value<'kvs>) -> Result<(), kv::Error> {
+//!         self.0.insert(key, value);
+//!
+//!         Ok(())
+//!     }
+//! }
+//!
+//! let mut visitor = Collect(BTreeMap::new());
+//!
+//! // info!(a = 1, b = 2, c = 3; "Something of interest");
+//!
+//! record.key_values().visit(&mut visitor)?;
+//!
+//! let collected = visitor.0;
+//!
+//! assert_eq!(
+//!     vec!["a", "b", "c"],
+//!     collected
+//!         .keys()
+//!         .map(|k| k.as_str())
+//!         .collect::<Vec<_>>(),
+//! );
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! [`Value`]s have methods for conversions to common types:
+//!
+//! ```
+//! # fn main() -> Result<(), log::kv::Error> {
+//! use log::kv::{Source, Key};
+//! # let record = log::Record::builder().key_values(&[("a", 1)]).build();
+//!
+//! // info!(a = 1; "Something of interest");
+//!
+//! let a = record.key_values().get(Key::from("a")).unwrap();
+//!
+//! assert_eq!(1, a.to_i64().unwrap());
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! Values also have their own [`VisitValue`] type. Value visitors are a lightweight
+//! API for working with primitives types:
+//!
+//! ```
+//! # fn main() -> Result<(), log::kv::Error> {
+//! use log::kv::{self, Source, Key, VisitValue};
+//! # let record = log::Record::builder().key_values(&[("a", 1)]).build();
+//!
+//! struct IsNumeric(bool);
+//!
+//! impl<'kvs> VisitValue<'kvs> for IsNumeric {
+//!     fn visit_any(&mut self, _value: kv::Value) -> Result<(), kv::Error> {
+//!         self.0 = false;
+//!         Ok(())
+//!     }
+//!
+//!     fn visit_u64(&mut self, _value: u64) -> Result<(), kv::Error> {
+//!         self.0 = true;
+//!         Ok(())
+//!     }
+//!
+//!     fn visit_i64(&mut self, _value: i64) -> Result<(), kv::Error> {
+//!         self.0 = true;
+//!         Ok(())
+//!     }
+//!
+//!     fn visit_u128(&mut self, _value: u128) -> Result<(), kv::Error> {
+//!         self.0 = true;
+//!         Ok(())
+//!     }
+//!
+//!     fn visit_i128(&mut self, _value: i128) -> Result<(), kv::Error> {
+//!         self.0 = true;
+//!         Ok(())
+//!     }
+//!
+//!     fn visit_f64(&mut self, _value: f64) -> Result<(), kv::Error> {
+//!         self.0 = true;
+//!         Ok(())
+//!     }
+//! }
+//!
+//! // info!(a = 1; "Something of interest");
+//!
+//! let a = record.key_values().get(Key::from("a")).unwrap();
+//!
+//! let mut visitor = IsNumeric(false);
+//!
+//! a.visit(&mut visitor)?;
+//!
+//! let is_numeric = visitor.0;
+//!
+//! assert!(is_numeric);
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! To serialize a value to a format like JSON, you can also use either `serde` or `sval`:
+//!
+//! ```
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # #[cfg(feature = "serde")]
+//! # {
+//! # use log::kv::Key;
+//! # #[derive(serde::Serialize)] struct Data { a: i32, b: bool, c: &'static str }
```

Comment:
I think we'll want to show this in the example so that people understand they need to implement/derive `serde::Serialize` for log (and `serde_json::to_string`)